core/templates/pages/exploration-player-page/services/content-translation-manager.service.ts

```diff
@@ -0,0 +1,218 @@
+import { Schema } from 'services/schema-default-value.service';
+import { SchemaConstants } from
+  'components/forms/schema-based-editors/schema-constants';
+import INTERACTION_SPECS from 'pages/interaction-specs.constants.ajs';
+import { WrittenTranslations } from
+  'domain/exploration/WrittenTranslationsObjectFactory';
+import { SubtitledUnicode } from
+  'domain/exploration/SubtitledUnicodeObjectFactory';
+import { ExtensionTagAssemblerService } from
+  'services/extension-tag-assembler.service';
+
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ContentTranslationManagerService {
+  private explorationLanguageCode: string;
+  private onStateCardContentUpdateEmitter: EventEmitter<void> = (
+    new EventEmitter());
+  // The 'originalTranscript' is a copy of the transcript in the exploration
+  // language in it's initial state.
+  private originalTranscript: StateCard[] = [];
+
+  constructor(
+    private playerTranscriptService: PlayerTranscriptService,
+    private extensionTagAssemblerService: ExtensionTagAssemblerService
+  ) {}
+
+  init(explorationLanguageCode: string) : void {
+    this.explorationLanguageCode = explorationLanguageCode;
+    this.originalTranscript = cloneDeep(
+      this.playerTranscriptService.transcript);
+  }
+
+  get onStateCardContentUpdate(): EventEmitter<void> {
+    return this.onStateCardContentUpdateEmitter;
+  }
+
+  /**
+   * This method replaces the translatable content inside the player transcript
+   * service's StateCards. If the language code is set back to the original
+   * exploration language, the original transcript in the exploration language
+   * is restored, since it was previously modified from switching to another
+   * language previously. Note that learners can only freely switch content
+   * languages on the first state and when there are no response pairs,
+   * otherwise they will have to refresh the page and restart the exploration.
+   * @param {string} languageCode The language code to display translations for.
+   */
+  displayTranslations(languageCode: string) : void {
+    const cards = this.playerTranscriptService.transcript;
+
+    if (languageCode === this.explorationLanguageCode) {
+      this.playerTranscriptService.restoreImmutably(this.originalTranscript);
+    } else {
+      cards.forEach(
+        card => this._displayTranslationsForCard(card, languageCode));
+    }
+
+    this.onStateCardContentUpdateEmitter.emit();
+  }
+
+  _swapContent(
+      writtenTranslations: WrittenTranslations,
+      languageCode: string,
+      content: SubtitledHtml|SubtitledUnicode
+  ): void {
+    const writtenTranslation = writtenTranslations.translationsMapping[
+      content.contentId][languageCode];
+
+    if (!this._isValidTranslation(writtenTranslation)) {
+      return;
+    }
+
+    let valueName;
+    if (content instanceof SubtitledHtml) {
+      valueName = 'html';
+    } else if (content instanceof SubtitledUnicode) {
+      valueName = 'unicode';
+    }
```

Comment:
else raise

Reply:
Ah, missed that `content` has a TS definition permitting only these two types. Please instead add a comment above this if/else block saying "Note that `content` can only be of type SubtitledHtml|SubtitledUnicode".

Reply:
Done!